lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,240 @@
+        Puppet::StateMachine.new(@machine_name, @states, @start_state)
+      end
+
+      # Compose multiple state machines into a a single state machine.
+      #
+      # @example
+      #   machine.compose({machine1: m1, machine2: m2}) do |m|
+      #     m.start_state([:m1, :start])
+      #
+      #     m.transition(
+      #       source: [:m1, :final],
+      #       target: [:m2, :start],
+      #       on: :'*final*')
+      #
+      #     m.final([:m2, :final])
+      #   end
+      def compose(machines, &blk)
+        machines.each_pair do |ns, machine|
+          @states.merge!(machine.namespaced_states(ns))
+        end
+        build(&blk)
+      end
+
+      # Verify that the machine is well defined.
+      # @return [void]
+      def verify!
+        verify_start_state!
+        verify_transition_targets!
+        verify_exiting_transitions!
+      end
+
+      # Define the start state for this state machine.
+      # @param name [Symbol]
+      # @return [void]
+      def start_state(name)
+        @start_state = name
+      end
+
+      # Generate a new state.
+      #
+      # @example
+      #   machine.state(:create_lockfile,
+      #     action: -> { Puppet::Util::Lockfile.new("/var/lock/puppet.lock").lock },
+      #     event: ->(has_lock) { has_lock ? :got_lock : :already_locked },
+      #     transitions: {
+      #       got_lock: :run_agent,
+      #       already_locked: :cancel_run,
+      #     })
+      #
+      # @return [Puppet::StateMachine::State] The generated machine state
+      def state(state_name, opts = {})
+        unhandled = opts.keys - [:action, :event, :type, :transitions]
+        if !unhandled.empty?
+          raise ArgumentError, "Cannot define state #{state_name.inspect}: unhandled attributes #{unhandled.inspect}"
+        end
+        if @states[state_name]
+          raise ArgumentError, "State #{state_name.inspect} already defined for state machine #{@machine_name.inspect}"
+        end
+        if [:error, :final].include?(opts[:type])
+          opts[:action] ||= lambda { }
+          if opts[:event]
+            raise ArgumentError, "State #{state_name.inspect} with terminating type #{opts[:type]} cannot be defined with an event proc"
+          end
+        else
+          if !opts[:action]
+            raise ArgumentError, "State #{state_name.inspect} with must be defined with an action proc"
+          end
+          if !opts[:event]
+            raise ArgumentError, "State #{state_name.inspect} with must be defined with an event proc"
+          end
+        end
+        trans = {}
+        if opts[:transitions]
+          opts[:transitions].each do |event, target|
+            trans[event] = target
+          end
+        end
+        @states[state_name] = State.new(state_name, opts[:action], opts[:event], opts[:type], trans)
+      end
+
+      # Add a transition between states.
+      #
+      # @note The source state must be defined before defining transitions for that state.
```

Comment:
I thought I responded to this comment but it seems to have vanished. Anywho, transitions are defined as `{event => :state_name}`, so there's no reason this wouldn't work. That said I haven't had the need for self referencing transitions so this isn't under test. When this behavior is needed we can explore it further.